### DISCUSSION TITLE
Custom Driver for Jetson Nano

### DISCUSSION THREAD

**Discussion by VishwanathB45** [OE4T/meta-tegra#1344] [Q&A]
Hi,

I'm using **kirkstone-l4t-r32.7.x** branch and using the instructions from Wiki I was able to build and flash Jetson Nano using core-image-sato and it worked !.

Now I'm trying to customize/change the devicetree and add custom camera driver to the bitbake build. But I'm unable to achive this objective as I'm unable to find any references to make this happen.

Could you please provide me with some pointers or help me with advice for achieving this goal.

**Comment by joekale-pp** [comment]
The recipe that points to the kernel sources is https://github.com/OE4T/meta-tegra/blob/kirkstone-l4t-r32.7.x/recipes-kernel/linux/linux-tegra_4.9.bb. You would be able to fork the kernel and replace the SRC_URI information with your fork of the kernel. Otherwise you could use `devtool` within bitbake to allow local modifications.

**Comment by kekiefer** [comment]
For changes to the devicetree, have a look at [tegrademo-devicetree](https://github.com/OE4T/tegra-demo-distro/tree/master/layers/meta-tegrademo/recipes-bsp/tegrademo-devicetree). This way you don't have to carry around kernel patches between releases. You will set `PREFERRED_PROVIDER_virtual/dtb` to your devicetree recipe, unset `KERNEL_DEVICETREE` and set `DTBFILE` to the dtb you create here. Overlays for the bootloader can be shipped this way as well.

Kernel modules can be built out of tree according to the [Working with Out-of-Tree Modules](https://docs.yoctoproject.org/kernel-dev/common.html#working-with-out-of-tree-modules) in the yp documentation. Depending on what sort of integration a camera module needs, you can also consider patching the kernel as @joekale-pp suggests if that is easier.

**Reply by VishwanathB45** [reply]
Hi @kekiefer 
I have did the same as you said above on Working with Out-of-Tree Modules ,I was able to add the cameradriver to the jetson build.Now to add our own device tree files for the camera driver ,what are the steps do I need to follow?

**Reply by kekiefer** [reply]
You'll construct a recipe similar to the one I linked in your layer. Unfortunately, there's no example in tegrademo for the machine you're interested in. So your kernel include path will be a little different and you will need to [include the entries from linux-tegra-4.9](https://github.com/OE4T/linux-tegra-4.9/blob/oe4t-patches-l4t-r32.7.4/arch/arm64/boot/dts/Makefile#L41-L48). You will set `KERNEL_DEVICETREE = ""` and `DTBFILE` to the dtb you build with this recipe in your local.conf or machine.conf if you have a custom one.

**Reply by VishwanathB45** [reply]
I don't have a custom devicetree file.
I had downloaded the bsp sources from nvidia download center and had modified devicetree sources to configure our camera driver, by replacing the raspi camera dtsi include (i.e. #include "porg-platforms/tegra210-porg-camera-rbpcv3-dual-imx477.dtsi" ) with our own custom dtsi file(i.e. #include "porg-platforms/tegra210-porg-camera-apracam.dtsi") in tegra210-p3448-0000-p3449-0000-b00.dts.
Then we built the devicetree and deployed using nvidia's sdk manager where we saw our camera was able to get configured and working. 
Now the same thing I wanted to do with yocto in automated way, but unable to figure out where I need to add my changes(devicetree patches) using the recipe.
Because of limited documentation/references I'm unable to achieve this goal
It would be very helpful for me, if you could give me some instructions or sample on how to move forward.

**Reply by spb06** [reply]
Hi there. I am going through something similar with a custom carrier board (running AGX Orin) and need to swap out the AGX Orin Devkit's .dtb for the custom .dtb. Were you able to figure this out? Any guidance will be greatly appreciated, thanks!
